---
layout: post
title:  "Visualizing Tighter Clustering Using More PCs"
author: Kyra Bowden
jhed: kbowden5
categories: [ EC HW1 ]
image: homework/hw6/hw6_kbowden5.png
featured: false
---

## Description


```{r}
## code credit to the hands-on activity in class
## read in the data
data = read.csv('/Users/Kyra_1/Documents/College/Junior/Spring/Genomic_Data_Visualization/Homework_1/codex_spleen_subset.csv.gz', 
                row.names = 1)

head(data)
dim(data)

pos <- data[, 1:2]
area <- data[, 3]
pexp <- data[, 4:31]
head(pexp)

## import necessary libraries
library(ggplot2)
library(patchwork)
library(dplyr)

## plot x,y coordinates and visualize area
df <- data.frame(pos, area)
ggplot(df) + geom_point(aes(x=x, y=y, col=area), size=0.1)

## normalize the protein expression data
hist(log10(colSums(pexp)+1))
hist(log10(rowSums(pexp)+1))
plot(log10(area+1), log10(rowSums(pexp)+1), pch=".")

pexp.norm <- log10(pexp/rowSums(pexp) * mean(rowSums(pexp))+1)

hist(pexp.norm[,'CD4'])
hist(log10(pexp.norm[,'CD4']+1))

## kmeans clustering
tw <- sapply(1:30, function(i) {
  kmeans(pexp.norm, centers=i)$tot.withinss
})
plot(tw, type='l')

## dimensionality reduction
pvar <- apply(pexp.norm, 2, var)
pmean <- apply(pexp.norm, 2, mean)
sort(apply(pexp.norm, 2, var))

pcs <- prcomp(pexp.norm)
head(pcs$x)

plot(pcs$sdev[1:30])
emb <- Rtsne::Rtsne(pexp.norm)$Y
head(emb)


## get clusters
tw <- sapply(1:20, function(i) {
  print(i)
  kmeans(emb, centers=i)$tot.withinss
})
plot(tw, type='l')

# 6 is probably ideal, testing 2
com = as.factor(kmeans(emb, centers=5)$cluster)

# clustering in reduced dimensional space
p1=ggplot(data.frame(emb, com=com)) + geom_point(aes(x = X1, y = X2, col=com), size=0.1)+
  labs(title = "Clustering tSNE Embedding")+ theme_minimal()

# clustering in physical space
p2 = ggplot(data.frame(pos, com = com))+ geom_point(aes(x = x, y = y, col = com), size = 0.1)+
  labs(title = "Cluster Physical Orientation") + theme_minimal()


## most differentially expressed gene in each cluster
results2 <- sapply(1:5, function(i){
  print(i)
  results = sapply(colnames(pexp), function(g){
    wilcox.test(pexp.norm[com == i, g],
                pexp.norm[com != i, g], 
                alternative = "greater")$p.val
  })
  topgene = names(head(sort(results, decreasing= F))[1])
  print(topgene)
  return(topgene)
})

# cluster 1 most differentially expressed gene
ggplot(data.frame(emb, com=com)) + geom_point(aes(x = X1, y = X2, col=pexp.norm$CD68), size=0.1)+
  scale_color_viridis_c('CD68', option="C")+
  labs(title = "Cluster CD68")+theme_minimal()

# cluster 2+5 most differentially expressed gene
p3 = ggplot(data.frame(emb, com=com)) + geom_point(aes(x = X1, y = X2, col=pexp.norm$CD8), size=0.1)+
  scale_color_viridis_c('CD8', option="C")+
  labs(title = "Cluster CD8")+theme_minimal()
# spatial representation of this gene
p4 = ggplot(data.frame(pos, com=com)) + geom_point(aes(x = x, y = y, col=pexp.norm$CD8), size=0.1)+
  scale_color_viridis_c('CD8', option="C")+
  labs(title = "CD8 Spatial Orientation")+theme_minimal()

# cluster 3
p5 = ggplot(data.frame(emb, com=com)) + geom_point(aes(x = X1, y = X2, col=pexp.norm$SMActin), size=0.1)+
  scale_color_viridis_c('SMActin', option="C")+
  labs(title = "Cluster SMActin")+theme_minimal()
p6 = ggplot(data.frame(pos, com=com)) + geom_point(aes(x = x, y = y, col=pexp.norm$SMActin), size=0.1)+
  scale_color_viridis_c('SMActin', option = "C")+
  labs(title = "SMActin Spatial Orientation")+theme_minimal()

# cluster 4
p7 = ggplot(data.frame(emb, com=com)) + geom_point(aes(x = X1, y = X2, col=pexp.norm$CD15), size=0.1)+
  scale_color_viridis_c('CD15', option = "C")+
  labs(title = "Cluster CD15")+theme_minimal()

## calculating pvals and fold change
pvals <- sapply(colnames(pexp), function(p) {
  print(p)
  test <- t.test(pexp.norm[com == 2, p], pexp.norm[com != 2, p])
  test$p.value
})
fc <- sapply(colnames(pexp), function(p) {
  print(p)
  mean(pexp.norm[com == 2, p])/mean(pexp.norm[com != 2, p])
})
pvals
fc
df <- data.frame(pv = -log10(pvals), log2fc = log2(fc), label=colnames(pexp))
head(df)
ggplot(df) + geom_point(aes(x = log2fc, y = pv)) + 
  geom_text(aes(x = log2fc, y = pv, label=label))

## volcano plot
# Convert p-values and fold changes into a data frame
df <- data.frame(pv = -log10(pvals), log2fc = log2(fc), label = colnames(pexp))

# Add a column to classify genes
df$status <- ifelse(df$log2fc > 0.3 & df$pv > -log10(0.05), "Upregulated",
                    ifelse(df$log2fc < -0.3 & df$pv > -log10(0.05), "Downregulated", "Not significant"))

# Basic volcano plot with color coding
volcano_plot <- ggplot(df, aes(x = log2fc, y = pv, color = status)) +
  geom_point(alpha=0.5) +
  scale_color_manual(values = c("Upregulated" = "red", "Downregulated" = "blue", "Not significant" = "grey50")) +
  theme_minimal() +
  labs(x = "Log2 Fold Change", y = "-Log10 P-value", title = "Volcano Plot")

# Identifying the most differentially expressed genes to label
# You can adjust the thresholds for labeling as per your data
label_genes <- df %>% filter(pv > -log10(0.05) & (log2fc > 0.3 | log2fc < -0.3))

# Adding labels to the plot
p8 = volcano_plot + geom_text(data = label_genes, aes(label = label), vjust = 1.5, hjust = 0.5, size = 3, check_overlap = TRUE)


## put it all together with patchwork
row1 <- p1 + p3 + p5
row2 = p2 + p4 + p6 
top_rows = row1/row2

third_row <- p7 | p8 

# Combine the rows into one plot with specified heights and widths
final_plot <- top_rows / 
  third_row +
  plot_layout(heights = c(1, 1, 1), widths = c(1, 2))+plot_annotation(tag_levels = 'A')

print(final_plot)

```
